xdp-dump/xdpdump.c

```diff
@@ -1603,14 +1603,23 @@ static bool capture_on_interface(struct dumpopt *cfg)
 	struct capture_programs      tgt_progs = {};
 
 	mp = xdp_multiprog__get_from_ifindex(cfg->iface.ifindex);
-	if (IS_ERR_OR_NULL(mp)) {
+	if (IS_ERR_OR_NULL(mp) ||
+	    (!IS_ERR_OR_NULL(mp) && xdp_multiprog__main_prog(mp) == NULL)) {
+
+		if (!IS_ERR_OR_NULL(mp)) {
+			xdp_multiprog__close(mp);
+			mp = NULL;
+		}
+
 		if (!cfg->load_xdp) {
-			pr_warn("WARNING: Specified interface does not have an XDP "
-				"program loaded, capturing\n         in legacy mode!\n");
+			pr_warn("WARNING: Specified interface does not have an XDP program loaded%s,"
+				"\n         capturing in legacy mode!\n",
+				IS_ERR_OR_NULL(mp) ? "" : " in software");
```

Comment:
You just closed `mp` and set it to `NULL` above, so this `IS_ERR_OR_NULL()` will always be true...

Reply:
Messed this up, when I was adding the mp = NULL to fix another error corner case I ran into. Will fix it differently ;)